intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,518 @@
+                          .align_to(..., 'H', 'T', 'D_head'))
+
+        if self_attn:
+            key = value = query
+        elif value is None:
+            # key and value are the same, but query differs
+            key = key.refine_names(..., 'T', 'D')
+            value = key
+        key_len = key.size('T')
+        dim = key.size('D')
+
+        # Distinguish between query_len (T) and key_len (T_key) dims.
+        k = prepare_head(self.k_lin(key)).rename(T='T_key')
+        v = prepare_head(self.v_lin(value)).rename(T='T_key')
+        q = prepare_head(self.q_lin(query))
+
+        dot_prod = q.div_(scale).matmul(k.align_to(..., 'D_head', 'T_key'))
+        dot_prod.refine_names(..., 'H', 'T', 'T_key')  # just a check
+
+        # (III)
+        # Named tensors doesn't support `==` yet; the following is a workaround.
+        attn_mask = (mask.rename(None) == 0).refine_names(*mask.names)
+        attn_mask = attn_mask.align_as(dot_prod)
+        dot_prod.masked_fill_(attn_mask, -float(1e20))
+
+        attn_weights = self.attn_dropout(F.softmax(dot_prod / scale, dim='T_key'))
+
+        # (IV)
+        attentioned = (
+            attn_weights.matmul(v).refine_names(..., 'H', 'T', 'D_head')
+            .align_to(..., 'T', 'H', 'D_head')
+            .flatten(['H', 'D_head'], 'D')
+        )
+
+        return self.out_lin(attentioned).refine_names(..., 'T', 'D')
+
+######################################################################
+# **(I) Refining the input tensor dims**
+
+def forward(self, query, key=None, value=None, mask=None):
+    # (I)
+    query = query.refine_names(..., 'T', 'D')
+
+######################################################################
+# The ``query = query.refine_names(..., 'T', 'D')`` serves as enforcable documentation
+# and lifts input dimensions to being named. It checks that the last two dimensions
+# can be refined to `['T', 'D']`, preventing potentially silent or confusing size
+# mismatch errors later down the line.
+#
+# **(II)  Manipulating dimensions in ``prepare_head``**
+
+# (II)
+def prepare_head(tensor):
+    tensor = tensor.refine_names(..., 'T', 'D')
+    return (tensor.unflatten('D', [('H', n_heads), ('D_head', dim_per_head)])
+                  .align_to(..., 'H', 'T', 'D_head'))
+
+######################################################################
+# The first thing to note is how the code clearly states the input and
+# output dimensions: the input tensor must end with the `T` and `D` dims
+# and the output tensor ends in `H`, `T`, and `D_head` dims.
+#
+# The second thing to note is how clearly the code describes what is going on.
+# prepare_head takes the key, query, and value and splits the embedding dim into
+# multiple heads, finally rearranging the dim order to be `[..., 'H', 'T', 'D_head']`.
+# ParlAI implements prepare_head as the following, using `view` and `transpose`
+# operations:
+#
+# **(III) Explicit broadcasting by names**
+
+def prepare_head(tensor):

```

Comment:
Needs to be changed as per https://github.com/pytorch/tutorials/pull/676/files#r332304115